File: 0213-House Robber II/0213-House Robber II.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if (n == 0)
        {
            return 0;
        }
        else if (n == 1)
        {
            return nums[0];
        }

        return max(rob(nums, 0, n - 2), rob(nums, 1, n - 1));
    }

private:
    int rob(vector<int>& nums, int start, int end) {
        int dp0 = 0, dp1 = 0;
        for (int i = start; i <= end; ++i)
        {
            int dp2 = max(dp1, dp0 + nums[i]);
            dp0 = dp1;
            dp1 = dp2;
        }
        
        return dp1;
    }
};

File: 0201-0300/0213-House Robber II/0213-House Robber II.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        def rob(start, end):
            dp0 = dp1 = 0
            for i in range(start, end + 1):
                dp2 = max(dp1, dp0 + nums[i])
                dp0 = dp1
                dp1 = dp2
            return dp1
        
        if len(nums) == 0:
            return 0
        elif len(nums) == 1:
            return nums[0]
        else:
            return max(rob(0, len(nums) - 2), rob(1, len(nums) - 1))
